File: Sources/iTunesKit/Services/WebCatalogService.swift
import Foundation

/// A service for interacting with the Apple Music Catalog Web API.
public actor WebCatalogService {
    
    private let client: iTunesWebServiceClient
    
    public init(client: iTunesWebServiceClient) {
        self.client = client
    }
    
    /// Fetches song details including editorial video information.
    /// - Parameter songId: The Apple Music song ID.
    /// - Returns: An `iTunesCatalogResponse` object.
    public func fetchSongDetails(songId: String, storefront: String = "us") async throws -> iTunesCatalogResponse {
        let endpoint = "catalog/\(storefront)/songs/\(songId)"
        
        let queryItems = [
            URLQueryItem(name: "art[url]", value: "f"),
            URLQueryItem(name: "fields[albums]", value: "editorialVideo"),
            URLQueryItem(name: "fields[songs]", value: "name,url,hasLyrics"),
            URLQueryItem(name: "format[resources]", value: "map"),
            URLQueryItem(name: "include[songs]", value: "albums"),
            URLQueryItem(name: "l", value: "en-US"),
            URLQueryItem(name: "omit[resource]", value: "autos"),
            URLQueryItem(name: "platform", value: "web")
        ]
        
        return try await client.send(endpoint, queryItems: queryItems)
    }
    
    /// Searches for a song using the legacy iTunes Search API (safer).
    /// - Parameters:
    ///   - term: The search term.
    ///   - storefront: The country storefront (e.g., "us").
    /// - Returns: The Apple Music song ID.
    public func search(term: String, storefront: String = "us") async throws -> String? {
        let urlString = "\(Secrets.itunesBaseURL)/search"
        guard var components = URLComponents(string: urlString) else { return nil }
        
        components.queryItems = [
            URLQueryItem(name: "term", value: term),
            URLQueryItem(name: "country", value: storefront),
            URLQueryItem(name: "media", value: "music"),
            URLQueryItem(name: "limit", value: "1")
        ]
        
        guard let url = components.url else { return nil }
        
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(iTunesSearchResponse.self, from: data)
        
        if let trackId = response.results.first?.trackId {
            return "\(trackId)"
        }
        
        return nil
    }
}

File: ituneskit/src/main/java/aaravgupta/ituneskit/services/WebCatalogService.kt
package aaravgupta.ituneskit

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

public class WebCatalogService(
    private val client: iTunesWebServiceClient,
    private val session: OkHttpClient = NetworkDefaults.defaultOkHttpClient(),
    private val json: Json = NetworkDefaults.defaultJson
) {
    public suspend fun fetchSongDetails(
        songId: String,
        storefront: String = "us"
    ): iTunesCatalogResponse {
        val endpoint = "catalog/$storefront/songs/$songId"
        val queryItems = listOf(
            URLQueryItem("art[url]", "f"),
            URLQueryItem("fields[albums]", "editorialVideo"),
            URLQueryItem("fields[songs]", "name,url,hasLyrics"),
            URLQueryItem("format[resources]", "map"),
            URLQueryItem("include[songs]", "albums"),
            URLQueryItem("l", Secrets.defaultLanguage),
            URLQueryItem("omit[resource]", "autos"),
            URLQueryItem("platform", "web")
        )

        return client.send(endpoint = endpoint, queryItems = queryItems)
    }

    public suspend fun search(term: String, storefront: String = "us"): String? {
        val baseURL = "${Secrets.itunesBaseURL}/search"
        val parsed = baseURL.toHttpUrlOrNull() ?: throw iTunesKitException.BadUrl(baseURL)
        val url = parsed.newBuilder()
            .addQueryParameter("term", term)
            .addQueryParameter("country", storefront)
            .addQueryParameter("media", "music")
            .addQueryParameter("limit", "1")
            .build()

        val request = Request.Builder().url(url).get().build()

        val data = withContext(Dispatchers.IO) {
            session.newCall(request).execute().use { response ->
                response.body?.bytes() ?: ByteArray(0)
            }
        }

        return try {
            val payload = json.decodeFromString(iTunesSearchResponse.serializer(), data.toString(Charsets.UTF_8))
            payload.results.firstOrNull()?.trackId?.toString()
        } catch (error: SerializationException) {
            throw iTunesKitException.CannotParseResponse(error)
        }
    }
}